packages/xo-server/src/index.js

```diff
@@ -140,6 +148,54 @@ async function setUpPassport(express, xo) {
     res.redirect('/')
   })
 
+  express.get('/signin/otp', (req, res, next) => {
+    const { user } = req.session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    res.send(
+      signInPage({
+        error: req.flash('error')[0],
+        otp: true,
+        strategies,
+      })
+    )
+  })
+
+  express.post('/signin/otp', (req, res, next) => {
+    const { body, session } = req
+    const { user } = session
+
+    if (!user) {
+      return res.redirect('/signin')
+    }
+
+    const isValid = authenticator.check(body.otp, user?.preferences?.otp)
+    if (isValid) {
+      setToken(req, res, next)
+    } else {
+      req.flash('error', 'Invalid code')
+      return res.redirect('/signin/otp')
```

Comment:
A [Post/Redirect/Get](https://en.wikipedia.org/wiki/Post/Redirect/Get) should be done with a status code of 303.